### PR TITLE
feat: artifacts upload support

### LINKED ISSUE #40: [Feature Request]: hook the nx cloud workflow up with build artifacts
Hey guys, 

I would like to know if it's currently possible to hook the nx cloud main workflow up with build artifacts?
To put it simple, I use this:
```
parallel-commands-on-agents: |
        npx nx affected --target=lint --parallel=3
        npx nx affected --target=test --parallel=3 --ci --code-coverage
        npx nx affected --target=build --parallel=3
```

And I would like to upload the generated code coverage to codecov via another job.
The problem, I can't use the generated files from the main nx workflow because I did not find any way to upload the artifacts with a command like:

```
      - uses: actions/upload-artifact@master
        with:
          name: dist
          path: dist
```

Right now my workaround has been to copy paste all your script, and add a step after the parallel commands have been executed:
```
  - name: Upload coverage to Codecov
    uses: codecov/codecov-action@v3
```

Thank you in advance for your attention 😊

### PR DESCRIPTION
**Pull request by gperdomor**
Closes #40

### CODE REVIEW

**Review by JamesHenry — CHANGES_REQUESTED** [review]
Thanks @gperdomor! Please add an integration-test workflow which proves that the artifacts are there and can be downloaded in a subsequent job. You can look at the existing integration-test workflows for inspiration

**Comment by gperdomor** [comment]
@JamesHenry added... can you approve the workflow run?

**Comment by JamesHenry** [comment]
@gperdomor please don’t force push to a PR that is actively being reviewed. It forces the reviewer to start from scratch on every commit. The PR will be squash merged anyway so the commit history is not a concern

**Comment by gperdomor** [comment]
@JamesHenry no more commits required, the pipeline already pass in my fork

File: README.md [hunk 1]
```diff
@@ -205,6 +205,14 @@ jobs:
     # NOTE: If you change this option, make sure it matches the agent configuration
     # Default: ubuntu-latest
     runs-on: ""
+
+    # [OPTIONAL] If you want to upload artifacts, please provide the paths as are required by the
+    # [upload-artifact](https://github.com/actions/upload-artifact) action. The name of the artifacts
+    # will be `nx-main-artifacts`.
+    #
+    # NOTE: To download your artifact in anothe job you need to use the [download-artifact](https://github.com/actions/download-artifact) action
```
Comment (JamesHenry) [inline README.md]:
```suggestion
    # NOTE: To download your artifact in another job you need to use the [download-artifact](https://github.com/actions/download-artifact) action
```

**Review by JamesHenry — APPROVED** [review]

**Comment by meeroslav** [comment]
Hey @gperdomor did you by any chance test if this solution works with multiline strings (passing multiple paths)?

We can extend this test:
```yaml
parallel-commands-on-agents: |
  npx nx run react-app:build
artifacts: ./integration-tests/npm/dist/
      
# ...
      
- run: node -e "if(fs.existsSync('dist/apps/react-app')) console.log('Artifacts are here...'); else { throw new Error('Artifacts do not exist!');}"      
```

to use wildcards and multiple artifact paths:

```yaml
parallel-commands-on-agents: |
  npx nx run react-app:build
  echo "hello world" > ./integration-tests/npm/dist/hello.txt
  echo "another file" > ./integration-tests/npm/dist/another.txt
artifacts: |
  ./integration-tests/npm/dist/apps/*
  ./integration-tests/npm/dist/*.txt
  
# ...
  
- run: node -e "if(fs.existsSync('dist/apps/react-app') && fs.existsSync('dist/hello.txt') && fs.existsSync('dist/another.txt')) console.log('Artifacts are here...'); else { throw new Error('Artifacts do not exist!');}"  
```

**Comment by gperdomor** [comment]
@meeroslav honestly not, but the upload is handled by the upload-artifact action so should works

**Comment by LayZeeDK** [comment]
I suggest generating test coverage reports, adding them to artifacts and verifying that they are also available to the dependent job.

Command:
```powershell
npx nx run react-app:test --ci --coverage
```

It will probably output to
```
coverage/apps/react-app
```

**Comment by gperdomor** [comment]
@LayZeeDK @meeroslav done with dist and coverage folder :D

**Review by LayZeeDK — APPROVED** [review]
This will be helpful!

**Review by LayZeeDK — CHANGES_REQUESTED** [review]
Bump the version to indicate a feature addition and prepare a release.

**Comment by gperdomor** [comment]
@LayZeeDK done :D

File: README.md [hunk 2]
```diff
@@ -208,6 +208,14 @@ jobs:
...
     # Default: ubuntu-latest
     runs-on: ""
+
+    # [OPTIONAL] If you want to upload artifacts, please provide the paths as are required by the
+    # [upload-artifact](https://github.com/actions/upload-artifact) action. The name of the artifacts
+    # will be `nx-main-artifacts`.
+    #
+    # NOTE: To download your artifact in another job you need to use the [download-artifact](https://github.com/actions/download-artifact) action
+    # Default: ""
+    artifacts: ""
```
Comment (SVendittelli) [inline README.md]:
```yaml
    # [OPTIONAL] Provide the name of uploaded artifacts.
    #
    # NOTE: This input only has an effect if used with the `artifacts` input.
    # Default: "nx-main-artifacts"
    artifactsName: ""
```

File: .github/workflows/nx-cloud-main.yml [hunk 3]
```diff
@@ -319,3 +323,10 @@ jobs:
         # It's important that we always run this step, otherwise in the case of any failures in preceding non-Nx steps, the agents will keep running and waste billable minutes
         if: ${{ always() }}
         run: npx nx-cloud stop-all-agents
+
+      - name: Uploading artifacts
+        uses: actions/upload-artifact@v3
+        if: ${{ inputs.artifacts != '' }}
+        with:
+          name: nx-main-artifacts
```
Comment (SVendittelli) [inline .github/workflows/nx-cloud-main.yml]:
```suggestion
          name: ${{ inputs.artifactsName }}
```

File: .github/workflows/nx-cloud-main.yml [hunk 4]
```diff
@@ -49,6 +49,10 @@ on:
         required: false
         type: string
         default: ubuntu-latest
+      artifacts:
+        required: false
+        type: string
+        default: ""
```
Comment (SVendittelli) [inline .github/workflows/nx-cloud-main.yml:55]:
```yaml
      artifactsName:
        required: false
        type: string
        default: nx-main-artifacts
```

**Review by SVendittelli — COMMENTED** [review]
Could we also provide an input for the name of the artefact? See individual comments for suggested implementation.

**Comment by gperdomor** [comment]
@SVendittelli done 👍🏻

File: .github/workflows/ci-integration-test-artifacts-multiple-folders.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    uses: ./.github/workflows/nx-cloud-main.yml
+    with:
+      working-directory: ./integration-tests/npm
+      parallel-commands-on-agents: |
+        npx nx run react-app:test --ci --coverage
+        npx nx run react-app:build
+      artifacts: |
+        ./integration-tests/npm/dist/
+        ./integration-tests/npm/coverage/
+      artifactsName: custom-artifacts-name
```
Comment (JamesHenry) [inline .github/workflows/ci-integration-test-artifacts-multiple-folders.yml]: We use `dash-case` for all other inputs, please align i.e. `artifacts-name`.

Additionally, if there is agreement that `artifacts-name` is going to be supported, then I wold vote to change `artifacts` to `artifacts-path` to make it clearer what it corresponds to

**Review by JamesHenry — CHANGES_REQUESTED** [review]
See comment

File: README.md [hunk 6]
```diff
@@ -208,6 +208,20 @@ jobs:
...
+    # [upload-artifact](https://github.com/actions/upload-artifact) action. The name of the artifacts
+    # will be `nx-main-artifacts`.
+    #
+    # NOTE: To download your artifact in another job you need to use the [download-artifact](https://github.com/actions/download-artifact) action
+    # Default: ""
+    artifacts-path: ""
+
+    # [OPTIONAL] Provide the name of uploaded artifacts.
+    #
+    # NOTE: This input only has an effect if used with the `artifacts` input.
```
Comment (JamesHenry) [inline README.md]:
```suggestion
    # NOTE: This input only has an effect if used with the `artifacts-path` input.
```

File: README.md [hunk 7]
```diff
@@ -208,6 +208,20 @@ jobs:
     # NOTE: If you change this option, make sure it matches the agent configuration
     # Default: ubuntu-latest
     runs-on: ""
+
+    # [OPTIONAL] If you want to upload artifacts, please provide the paths as are required by the
+    # [upload-artifact](https://github.com/actions/upload-artifact) action. The name of the artifacts
+    # will be `nx-main-artifacts`.
```
Comment (JamesHenry) [inline README.md]:
```suggestion
    # will be the value of the `artifacts-name` input.
```

**Review by JamesHenry — APPROVED** [review]
Thanks again @gperdomor!

**Comment by gperdomor** [comment]
Happy to help :D...

**Comment by meeroslav** [comment]
Thank you all for participating in this amazing new feature!

This will be super helpful for a lot of folks.

**Comment by bdebon** [comment]
Guys, I'm so immensely happy and grateful that you implemented so quickly this thing I was asking for. And it's wonderful to see this PR with so many people interacting to deliver a bulletproof feature.
Thanks for all!

**Pull request merged**